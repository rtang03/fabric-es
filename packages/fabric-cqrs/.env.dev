CHANNEL_NAME=loanapp
CA_NAME=rca-org1
CA_ENROLLMENT_ID_ADMIN=rca-org1-admin
CA_ENROLLMENT_SECRET_ADMIN=rca-org1-adminPW
ORDERER_HOSTNAME=orderer0-org0
ORDERER_TLSCA_CERT=../../deployments/dev-net/artifacts/crypto-config/Org0MSP/orderer0.org0.com/tls-msp/tlscacerts/tls-0-0-0-0-5052.pem
CONNECTION_PROFILE=connection/connection-org1.local.yaml
MSPID=Org1MSP
ORG_ADMIN_ID=admin-org1.net
ORG_ADMIN_SECRET=admin-org1.netPW
WALLET=assets/wallet

# Org2 (pbctfp) - SOLELY used for running service.multiusers.test.ts
ORG2_ORG_ADMIN_ID=admin-org2.net
ORG2_ORG_ADMIN_SECRET=admin-org2.netPW
ORG2_MSPID=Org2MSP
ORG2_CONNECTION_PROFILE=connection/connection-org2.local.yaml
ORG1_WALLET=assets/walletOrg1
ORG2_WALLET=assets/walletOrg2
ORG2_CA_NAME=rca-org1
ORG2_CA_ENROLLMENT_ID_ADMIN=rca-org2-admin
ORG2_CA_ENROLLMENT_SECRET_ADMIN=rca-org2-adminPW



