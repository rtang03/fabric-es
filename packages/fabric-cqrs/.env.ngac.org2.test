## unsure if this is required
COLLECTION=Org1PrivateDetails
PATH_TO_CHANNEL_CONFIG=../../artifacts/channel-config

### MANDATORY
#Orgnaization Admin EnrollmentID
ORG_ADMIN_ID=Admin@org2.example.com
#Organization Admin EnrollSecret
ORG_ADMIN_SECRET=peer1pw
#URL for Orgnization CA
ORG_CA_URL=https://0.0.0.0:5055
# org name defined in connnection profile; case sensitive
ORGNAME=Org2
# Path to Fabric Network
NETWORK_LOCATION=../../network/hosts
# Path to Connection Path
CONNECTION_PROFILE=connection/peer0org2.yaml
# Relative path to wallet directory
WALLET=assets/wallet
# Channel Event Hub
CHANNEL_HUB=peer0.org2.example.com
# Optional:
CHAINCODE_ID=eventstore
CHAINCODE_ID_PRIVATEDATA=privatedata
# Used in logger
APPLICATION=peer0org2app


