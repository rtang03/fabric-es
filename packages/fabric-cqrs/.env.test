#MSPID=Org1MSP
#CHANNEL_NAME=mychannel

COLLECTION=Org1PrivateDetails
PATH_TO_CHANNEL_CONFIG=../../artifacts/channel-config
CHANNEL_NAME=eventchannel

### MANDATORY
#Orgnaization Admin EnrollmentID
ORG_ADMIN_ID=Admin@org1.example.com
#Organization Admin EnrollSecret
ORG_ADMIN_SECRET=peer1pw
#URL for Orgnization CA
ORG_CA_URL=https://0.0.0.0:5054
# org name defined in connnection profile; case sensitive
ORGNAME=Org1
# Path to Fabric Network
NETWORK_LOCATION=../../network/hosts
# Path to Connection Path
CONNECTION_PROFILE=connection/peer0org1.yaml
# Relative path to wallet directory
WALLET=assets/wallet
# Channel Event Hub
CHANNEL_HUB=peer0.org1.example.com
# Optional:
CHAINCODE_ID=eventstore
CHAINCODE_ID_PRIVATEDATA=privatedata


