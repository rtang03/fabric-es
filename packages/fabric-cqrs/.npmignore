*
!dist/
!dist/**/*
!package.json
!*.md
!LICENSE
!README.md
