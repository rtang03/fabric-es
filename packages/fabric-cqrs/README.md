# fabric-cqrs

## Installation

```shell script
yarn install @fabric-es/fabric-cqrs
```
