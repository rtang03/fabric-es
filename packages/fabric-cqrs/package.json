{
  "name": "@espresso/fabric-cqrs",
  "version": "1.0.0",
  "main": "dist/index.js",
  "typings": "dist/index.d.ts",
  "scripts": {
    "build": "yarn run clean && tsc",
    "build:prod": "yarn run clean && tsc -p ./tsconfig.prod.json",
    "package": "yarn run build:cleanup && yarn run cp:package.json && yarn run cp:dist",
    "build:cleanup": "rimraf ../../.build/packages/fabric-cqrs && mkdir -p ../../.build/packages/fabric-cqrs",
    "cp:package.json": "cp ./package.json ../../.build/packages/fabric-cqrs/",
    "cp:dist": "cp -R ./dist ../../.build/packages/fabric-cqrs/",
    "clean": "rimraf dist",
    "test:account": "jest account.unit.test",
    "test:service.integration": "jest service.integration.test",
    "test:service.multiusers": "rimraf assets/walletOrg1 && rimraf assets/walletOrg2 && jest service.multiusers.test",
    "test:service.privatedata.integration": "jest service.privatedata.integration.test",
    "test:service.ngac.crud.integration": "(broke) jest service.ngac.crud.integration.test",
    "test:service.ngac.pe.int.1u": "(broke) jest one-user.service.ngac.pe.int.test",
    "test:store.integration": "jest command.integration.test",
    "test:store.privatedata.integration": "jest command.privatedata.integration.test",
    "test:peer.integration": "jest peer.integration.test",
    "test:peer.privatedata.integration": "jest peer.privatedata.integration.test",
    "test:peer.ngac.integration": "(not implemented yet) jest peer.ngac.integration.test",
    "test:all-unit-test": "jest unit.test",
    "enrollAdmin": "ts-node src/account/enrollAdmin.ts",
    "enrollCaAdmin": "ts-node src/account/enrollCaAdmin.ts",
    "enrollAdminOrg2": "(broke) ts-node src/ngac_test/enrollAdmin.ts",
    "typedoc": "typedoc --excludeNotExported --theme markdown --exclude \"**/*+(.test|constants).ts\" --out docs src/index.ts"
  },
  "engines": {
    "node": ">=8",
    "npm": ">=5"
  },
  "private": true,
  "engineStrict": true,
  "dependencies": {
    "@espresso/operator": "1.0.0",
    "dotenv": "^8.1.0",
    "fabric-client": "^1.4.4",
    "fabric-contract-api": "^1.4.2",
    "fabric-network": "^1.4.4",
    "js-yaml": "^3.13.1",
    "lodash": "^4.17.15",
    "prettyjson": "^1.2.1",
    "redux": "^4.0.4",
    "redux-observable": "^1.2.0",
    "rxjs": "^6.5.3",
    "winston": "^3.2.1"
  },
  "devDependencies": {
    "@types/bytebuffer": "^5.0.40",
    "@types/dotenv": "^6.1.1",
    "@types/jest": "^24.0.18",
    "@types/js-yaml": "^3.12.1",
    "@types/lodash": "^4.14.138",
    "@types/node": "^8.10.54",
    "jest": "^24.9.0",
    "prettier": "^1.19.1",
    "rimraf": "^3.0.0",
    "ts-jest": "^24.0.2",
    "ts-node": "^8.3.0",
    "tslint": "^5.20.0",
    "tslint-config-prettier": "^1.18.0",
    "tslint-plugin-prettier": "^2.0.1",
    "typedoc": "^0.15.7",
    "typedoc-plugin-markdown": "^2.2.16",
    "typescript": "^3.7.2"
  }
}
