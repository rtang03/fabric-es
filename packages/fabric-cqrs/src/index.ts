/**
 * @packageDocumentation
 * @hidden
 */
export {
  createPeer,
  getMockRepository,
  getPrivatedataMockRepository,
  createProjectionDb,
  createQueryDatabase
} from './peer';
export { getContract, getNetwork, submit, evaluate, evaluate$ } from './services';
export * from './types';
export {
  dispatchResult,
  generateToken,
  getAction,
  getErrorAction,
  getErrorActionHandler,
  getSuccessAction,
  getSuccessActionHandler
} from './cqrs/utils';
export { fromCommitsToGroupByEntityId, getHistory, isCommit } from './peer/utils';
