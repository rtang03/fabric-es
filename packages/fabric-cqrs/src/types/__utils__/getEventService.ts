import { Channel, EventListener, EventService, Eventer, Utils } from 'fabric-common';

export const getEventService: (option: {
  name: string;
  channel: Channel;
  onChannelEventArrived;
}) => {
  getService: () => EventService;
  send: () => any;
  register: () => void;
  unregister: () => void;
} = ({ name, channel, onChannelEventArrived }) => {
  Utils.getLogger('[fabric-cqrs] eventService.js');

  const service: EventService = channel.newEventService(name);
  let listener: EventListener;


  return {
    getService: () => service,
    send: async () => {
      // service.setTargets([{}]);
      // await service.send({} );
    },
    register: () => {
      listener = service.registerChaincodeListener(
        'eventstore',
        (error, event) => {
          console.log('--------------');
          console.log(event);
          const ccEvent = event.chaincodeEvents[0].payload.toString('utf8');
          console.log(ccEvent);
        },
        { unregister: false }
      );
    },
    unregister: () => listener.unregisterEventListener()
  };
};
