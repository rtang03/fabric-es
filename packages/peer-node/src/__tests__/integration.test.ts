require('dotenv').config();
import {
  Document,
  DocumentEvent,
  reduceToDocument,
  reduceToTrade,
  reduceToUser,
  Trade,
  TradeEvent,
  User,
  UserEvent
} from '@espresso/common';
import { createPeer, getNetwork, Peer } from '@espresso/fabric-cqrs';
import { ApolloServer } from 'apollo-server';
import { createTestClient } from 'apollo-server-testing';
import { omit, pick } from 'lodash';
import {
  CREATE_DOCUMENT,
  DOCUMENT_ETCPO_BY_ID,
  resolvers as docResolvers,
  typeDefs as docTypeDefs
} from '../document';
import {
  CREATE_ETCPO,
  EtcPo,
  ETCPO_BY_ID,
  EtcPoEvent,
  reduceToEtcPo,
  resolvers as privatedataResolvers,
  typeDefs as privatedataTypeDefs
} from '../privatedata';
import {
  CREATE_TRADE,
  resolvers as tradeResolvers,
  typeDefs as tradeTypeDefs
} from '../trade';
import {
  CREATE_USER,
  PAGINATED_USER,
  USER_BY_ID,
  USER_COMMITS,
  USERS
} from '../user';
import { constructTestServer, getApolloServer } from './__utils__';

let server;
let docPeer: Peer;
let tradePeer: Peer;
let privatePeer: Peer;
let documentService: ApolloServer;
let tradeService: ApolloServer;
let privatedataService: ApolloServer;
let docNetworkConfig;
let tradeNetworkConfig;
let privateNetworkConfig;
const collection = 'Org1PrivateDetails';
const prefix = 'int_test_';

beforeAll(async () => {
  // Document Service

  const enrollmentId = '';
  docNetworkConfig = await getNetwork({ enrollmentId });
  docPeer = createPeer({
    ...docNetworkConfig,
    collection,
    reducer: reduceToDocument
  });
  await docPeer.subscribeHub();
  documentService = getApolloServer({
    typeDefs: docTypeDefs,
    resolvers: docResolvers,
    dataSources: () => ({
      docDataSource: {
        repo: docPeer.getRepository<Document, DocumentEvent>({
          entityName: prefix + 'document',
          reducer: reduceToDocument
        })
      },
      userDataSource: {
        repo: docPeer.getRepository<User, UserEvent>({
          entityName: prefix + 'user',
          reducer: reduceToUser
        })
      },
      tradeDataSource: {
        repo: docPeer.getRepository<Trade, TradeEvent>({
          entityName: prefix + 'trade',
          reducer: reduceToTrade
        })
      }
    })
  });
  await documentService.listen({ port: 14001 });

  // Trade Service
  // todo: bug
  tradeNetworkConfig = await getNetwork({ enrollmentId: null });
  tradePeer = createPeer({
    ...tradeNetworkConfig,
    collection,
    reducer: reduceToDocument
  });
  await tradePeer.subscribeHub();
  tradeService = getApolloServer({
    typeDefs: tradeTypeDefs,
    resolvers: tradeResolvers,
    dataSources: () => ({
      userDataSource: {
        repo: tradePeer.getRepository<User, UserEvent>({
          entityName: prefix + 'user',
          reducer: reduceToUser
        })
      },
      tradeDataSource: {
        repo: tradePeer.getRepository<Trade, TradeEvent>({
          entityName: prefix + 'trade',
          reducer: reduceToTrade
        })
      }
    })
  });
  await tradeService.listen({ port: 14002 });

  // Privatedata Service
  privateNetworkConfig = await getNetwork({ enrollmentId });
  privatePeer = createPeer({
    ...privateNetworkConfig,
    collection,
    reducer: reduceToEtcPo
  });
  privatedataService = getApolloServer({
    typeDefs: privatedataTypeDefs,
    resolvers: privatedataResolvers,
    dataSources: () => ({
      etcDataSource: {
        privatedataRepo: privatePeer.getPrivateDataRepo<EtcPo, EtcPoEvent>({
          entityName: prefix + 'etcPo',
          reducer: reduceToEtcPo
        })
      }
    })
  });
  await privatedataService.listen({ port: 14003 });

  server = await constructTestServer();
});

afterAll(async () => {
  docPeer.unsubscribeHub();
  tradePeer.unsubscribeHub();
  docPeer.disconnect();
  tradePeer.disconnect();
  privatePeer.disconnect();
  await documentService.stop();
  await tradeService.stop();
  await privatedataService.stop();
  await server.stop();
});

let commitId: string;

describe('User Entity: Integration Test', () => {
  it('should create user', async () =>
    createTestClient(server)
      .query({
        query: CREATE_USER,
        variables: {
          name: 'Create Test User',
          userId: 'test@gmail.com'
        }
      })
      .then(({ data: { createUser } }) => {
        commitId = createUser.commitId;
        expect(pick(createUser, 'id', 'entityName')).toMatchSnapshot();
      }));

  it('should run QUERIES', done =>
    setTimeout(async () => {
      await createTestClient(server)
        .query({ query: USERS })
        .then(({ data }) => expect(data).toMatchSnapshot());

      await createTestClient(server)
        .query({
          query: PAGINATED_USER,
          variables: { cursor: 10 }
        })
        .then(({ data }) => expect(data).toMatchSnapshot());

      await createTestClient(server)
        .query({
          query: USER_BY_ID,
          variables: { id: 'test@gmail.com' }
        })
        .then(({ data }) => expect(data).toMatchSnapshot());

      await createTestClient(server)
        .query({
          query: USER_COMMITS,
          variables: { id: 'test@gmail.com' }
        })
        .then(({ data: { getCommitByUserId } }) => getCommitByUserId)
        .then(commits => omit(commits[0], 'commitId', 'committedAt'))
        .then(result => expect(result).toMatchSnapshot());

      done();
    }, 5000));
});

describe('Trade Entity: Integration Test', () => {
  it('should create trade', async () =>
    createTestClient(server)
      .query({
        query: CREATE_TRADE,
        variables: {
          tradeId: '123123123',
          userId: 'test@gmail.com',
          title: 'test-title',
          description: 'test-description'
        }
      })
      .then(({ data: { createTrade } }) => createTrade)
      .then(commit => pick(commit, 'id', 'entityName', 'version'))
      .then(result => expect(result).toMatchSnapshot()));
});

describe('Document Entity: Integration Test', () => {
  it('should create document', async () =>
    createTestClient(server)
      .query({
        query: CREATE_DOCUMENT,
        variables: {
          documentId: '321321321',
          tradeId: '123123123',
          userId: 'test@gmail.com',
          title: 'test-title',
          description: 'test-description',
          link: 'test-link'
        }
      })
      .then(({ data: { createDocument } }) => createDocument)
      .then(commit => pick(commit, 'id', 'entityName', 'version'))
      .then(result => expect(result).toMatchSnapshot()));
});

describe('Etc Privatedata: Integration Test', () => {
  it('should create EtcPo', async () =>
    createTestClient(server)
      .query({
        query: CREATE_ETCPO,
        variables: {
          id: '321321321',
          userId: 'test@gmail.com',
          body: 'etc po details'
        }
      })
      .then(({ data: { createEtcPo } }) => createEtcPo)
      .then(commit => pick(commit, 'id', 'entityName', 'version'))
      .then(result => expect(result).toMatchSnapshot()));

  it('should query: getEtcPoById', async () =>
    createTestClient(server)
      .query({ query: ETCPO_BY_ID, variables: { id: '321321321' } })
      .then(({ data }) => expect(data).toMatchSnapshot()));

  it('should run federated query: getDocumentEtcById', async () =>
    createTestClient(server)
      .query({ query: DOCUMENT_ETCPO_BY_ID, variables: { id: '321321321' } })
      .then(({ data }) => expect(data).toMatchSnapshot()));
});
