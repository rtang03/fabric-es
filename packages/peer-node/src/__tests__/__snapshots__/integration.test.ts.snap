// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Doc Contents Integration Test query doc content again by ID 1`] = `
Object {
  "body": "{ \\"id\\": \\"doc0009\\", \\"type\\": \\"Bill Lay Ding\\", \\"ref\\": \\"ref0009\\" }",
}
`;

exports[`Doc Contents Integration Test query doc content by ID 1`] = `
Object {
  "format": "PDF",
  "link": "localhost/docontent0008",
}
`;

exports[`Document Integration Test query commits of document 1`] = `
Array [
  Object {
    "entityName": "int_test_document",
    "events": Array [
      Object {
        "type": "DocumentCreated",
      },
      Object {
        "type": "DocumentReferenceDefined",
      },
      Object {
        "type": "DocumentLoanIdDefined",
      },
      Object {
        "type": "DocumentTitleDefined",
      },
    ],
  },
]
`;

exports[`Document Integration Test query eocument by ID 1`] = `
Object {
  "reference": "DOC0101",
  "status": 0,
  "title": "Test Title 101",
}
`;

exports[`Document Integration Test query focument by ID 1`] = `
Object {
  "reference": "DOC0102",
  "status": 0,
  "title": "Test Title 102",
}
`;

exports[`Document Integration Test query gocument by ID 1`] = `
Object {
  "reference": "DOC0103",
  "status": 0,
  "title": "Test Title 103",
}
`;

exports[`Federated queries federated query doc contents by ID 1`] = `
Object {
  "content": Object {
    "format": "PDF",
    "link": "localhost/docontent0008",
  },
  "document": Object {
    "reference": "DOC0102",
    "status": 1,
    "title": "Test Title 102",
  },
}
`;

exports[`Federated queries federated query document by ID 1`] = `
Object {
  "contents": Object {
    "content": Object {
      "format": "PDF",
      "link": "localhost/docontent0008",
    },
  },
  "loan": Object {
    "description": "loan-description-1",
    "reference": "MYLOAN0001",
    "status": 0,
  },
  "reference": "DOC0102",
  "status": 1,
  "title": "Test Title 102",
}
`;

exports[`Federated queries federated query loan by ID 1`] = `
Object {
  "description": "loan-description-1",
  "details": Object {
    "contact": Object {
      "contactName": "Hook",
      "email": "hook@fake.it",
      "phone": "555-23456",
    },
    "currency": "HKD",
    "loanType": "Post-shipment",
    "requestedAmt": 70000,
    "requester": Object {
      "companyName": "Peter Hook Ltd",
      "registration": "BR000000X",
    },
    "startDate": "2019-10-12",
    "tenor": 60,
  },
  "documents": Array [
    Object {
      "reference": "DOC0101",
      "status": 0,
      "title": "Test Title 101",
    },
    Object {
      "reference": "DOC0102",
      "status": 1,
      "title": "Test Title 102",
    },
  ],
  "reference": "MYLOAN0001",
  "status": 0,
}
`;

exports[`Federated queries federated query loan details by ID 1`] = `
Object {
  "approvedAmt": null,
  "comment": null,
  "contact": Object {
    "email": "pete@fake.it",
    "name": "Pete",
    "phone": "555-12345",
    "salutation": "Mr.",
    "title": "Owner",
  },
  "currency": "HKD",
  "loan": Object {
    "description": "test-description",
    "reference": "MYLOAN0001",
    "status": 0,
  },
  "loanType": "Post-shipment",
  "requestedAmt": 50000,
  "requester": Object {
    "name": "Pete N Co. Ltd",
    "registration": "BR0000001",
    "type": "Money Launderer",
  },
  "startDate": "2019-10-11",
  "tenor": 60,
}
`;

exports[`Loan Details Integration Test query loan details again by ID 1`] = `
Object {
  "approvedAmt": null,
  "comment": null,
  "contact": Object {
    "email": "crime@fake.it",
    "name": "Locke",
    "phone": "555-33457",
    "salutation": "Mr.",
    "title": "Owner",
  },
  "currency": "HKD",
  "loanType": "Post-shipment",
  "requestedAmt": 210000,
  "requester": Object {
    "name": "John Locke Ltd",
    "registration": "BR000000Y",
    "type": "Gangster",
  },
  "startDate": "2019-10-13",
  "tenor": 77,
}
`;

exports[`Loan Details Integration Test query loan details by ID 1`] = `
Object {
  "approvedAmt": null,
  "comment": null,
  "contact": Object {
    "email": "hook@fake.it",
    "name": "Hook",
    "phone": "555-23456",
    "salutation": "Mr.",
    "title": "Owner",
  },
  "currency": "HKD",
  "loanType": "Post-shipment",
  "requestedAmt": 70000,
  "requester": Object {
    "name": "Peter Hook Ltd",
    "registration": "BR000000X",
    "type": "Money Launderer",
  },
  "startDate": "2019-10-12",
  "tenor": 60,
}
`;

exports[`Loan Integration Test query commits of loan 1`] = `
Array [
  Object {
    "entityName": "int_test_loan",
    "events": Array [
      Object {
        "type": "LoanApplied",
      },
      Object {
        "type": "LoanReferenceDefined",
      },
      Object {
        "type": "LoanDescriptionDefined",
      },
    ],
  },
]
`;

exports[`Loan Integration Test query loan by ID 1`] = `
Object {
  "description": "loan-description-1",
  "reference": "MYLOAN0001",
  "status": 0,
}
`;

exports[`User Integration Test query commits of user 1`] = `
Array [
  Object {
    "entityName": "int_test_user",
    "events": Array [
      Object {
        "type": "UserCreated",
      },
    ],
  },
]
`;

exports[`User Integration Test query paginated user 1`] = `
Object {
  "getPaginatedUser": Object {
    "entities": Array [
      "The new User 2",
      "The new User 3",
      "The new User",
    ],
    "hasMore": false,
    "total": 3,
  },
}
`;
